File: frontend/src/app/core/notificaciones/notificaciones.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable, effect, inject, signal, untracked } from '@angular/core';
import { Observable, tap } from 'rxjs';

import { environment } from '../../../environments/environment';
import { AuthService } from '../auth/auth.service';
import {
  CantidadNoLeidasOut,
  LeerTodasOut,
  NotificacionOut,
  NotificacionPaginadoOut,
} from './notificaciones.models';

/** Cada cuanto se reconsulta el contador de la campana. No hay websockets: es polling simple. */
const INTERVALO_POLLING_MS = 60_000;

/**
 * Notificaciones del usuario de la sesion (PENDIENTES.txt 2.19.3), sea cliente o personal.
 *
 * `noLeidas` es el numero del badge de la campana. El polling arranca solo cuando hay sesion y
 * se detiene al cerrarla: lo maneja un `effect` sobre el usuario de `AuthService`, asi que ni la
 * cabecera ni el logout tienen que acordarse de apagarlo. Si cambia el usuario (logout y login
 * con otra cuenta) se reinicia y el contador vuelve a cero antes de la primera consulta.
 */
@Injectable({ providedIn: 'root' })
export class NotificacionesService {
  private readonly http = inject(HttpClient);
  private readonly auth = inject(AuthService);
  private readonly base = `${environment.apiUrl}/notificaciones`;

  readonly noLeidas = signal(0);

  private temporizador: ReturnType<typeof setInterval> | undefined;

  constructor() {
    effect(() => {
      const usuarioId = this.auth.usuario()?.id ?? null;
      untracked(() => {
        this.detenerPolling();
        this.noLeidas.set(0);
        if (usuarioId) this.iniciarPolling();
      });
    });
  }

  listar(pagina = 1, tamanioPagina = 20, soloNoLeidas = false): Observable<NotificacionPaginadoOut> {
    const params = new HttpParams()
      .set('pagina', pagina)
      .set('tamanio_pagina', tamanioPagina)
      .set('solo_no_leidas', soloNoLeidas);
    return this.http
      .get<NotificacionPaginadoOut>(this.base, { params })
      .pipe(tap((respuesta) => this.noLeidas.set(respuesta.no_leidas)));
  }

  marcarLeida(id: string): Observable<NotificacionOut> {
    return this.http.post<NotificacionOut>(`${this.base}/${id}/leida`, {}).pipe(
      // el contador exacto llega en el proximo refresco; mientras tanto se descuenta local
      tap(() => this.noLeidas.update((n) => Math.max(0, n - 1))),
    );
  }

  leerTodas(): Observable<LeerTodasOut> {
    return this.http
      .post<LeerTodasOut>(`${this.base}/leer-todas`, {})
      .pipe(tap(() => this.noLeidas.set(0)));
  }

  /** Reconsulta el contador; los errores se ignoran (la campana no es critica para la pagina). */
  refrescarContador(): void {
    if (!this.auth.estaAutenticado()) return;
    this.http.get<CantidadNoLeidasOut>(`${this.base}/no-leidas/cantidad`).subscribe({
      next: (respuesta) => this.noLeidas.set(respuesta.no_leidas),
      error: () => undefined,
    });
  }

  private iniciarPolling(): void {
    this.refrescarContador();
    this.temporizador = setInterval(() => {
      // con la pestania en segundo plano no vale la pena pegarle a la API
      if (typeof document !== 'undefined' && document.visibilityState === 'hidden') return;
      this.refrescarContador();
    }, INTERVALO_POLLING_MS);
  }

  private detenerPolling(): void {
    clearInterval(this.temporizador);
    this.temporizador = undefined;
  }
}

File: movil/lib/core/notificaciones/notificaciones_service.dart
import 'dart:async';

import 'package:flutter/widgets.dart';

import '../api.dart';
import '../auth/auth_service.dart';
import 'notificaciones_models.dart';

/// Notificaciones del usuario de la sesion (PENDIENTES.txt 2.19.3), espejo de
/// `NotificacionesService` del frontend. Es un ChangeNotifier porque el contador de no
/// leidas se muestra en el badge de la campana (tienda y cuenta).
///
/// Escucha a [AuthService]: con sesion arranca un polling cada 60 s del contador, y al cerrar
/// sesion (o si la API responde 401 y la sesion expira) lo detiene y vuelve a cero. Con la app
/// en segundo plano el tick se saltea.
class NotificacionesService extends ChangeNotifier {
  NotificacionesService(this._auth) {
    _auth.addListener(_alCambiarSesion);
    _alCambiarSesion();
  }

  static const Duration _intervalo = Duration(seconds: 60);

  final AuthService _auth;
  Timer? _temporizador;
  String? _usuarioId;

  int _noLeidas = 0;
  int get noLeidas => _noLeidas;

  void _alCambiarSesion() {
    final usuarioId = _auth.estaAutenticado ? _auth.usuario?.id : null;
    if (usuarioId == _usuarioId) return;
    _usuarioId = usuarioId;

    _temporizador?.cancel();
    _temporizador = null;
    _fijar(0);
    if (usuarioId == null) return;

    refrescarContador();
    _temporizador = Timer.periodic(_intervalo, (_) {
      final estado = WidgetsBinding.instance.lifecycleState;
      if (estado != null && estado != AppLifecycleState.resumed) return;
      refrescarContador();
    });
  }

  void _fijar(int valor) {
    if (valor == _noLeidas) return;
    _noLeidas = valor;
    notifyListeners();
  }

  /// Reconsulta el contador; los errores se ignoran (la campana no es critica).
  Future<void> refrescarContador() async {
    if (!_auth.estaAutenticado) return;
    try {
      final respuesta = await api.get('/notificaciones/no-leidas/cantidad');
      // la sesion pudo cerrarse mientras volvia la respuesta
      if (!_auth.estaAutenticado) return;
      _fijar(aEntero((respuesta as Map<String, dynamic>)['no_leidas']));
    } catch (_) {}
  }

  Future<NotificacionPaginadoOut> listar({
    int pagina = 1,
    int tamanioPagina = 20,
    bool soloNoLeidas = false,
  }) async {
    final respuesta = await api.get('/notificaciones', query: {
      'pagina': pagina,
      'tamanio_pagina': tamanioPagina,
      'solo_no_leidas': soloNoLeidas,
    });
    final resultado = NotificacionPaginadoOut.desdeJson(respuesta as Map<String, dynamic>);
    _fijar(resultado.noLeidas);
    return resultado;
  }

  /// Solo tiene sentido para una notificacion todavia no leida: descuenta el badge local
  /// y el valor exacto llega en el proximo refresco.
  Future<NotificacionOut> marcarLeida(String id) async {
    final respuesta = await api.post('/notificaciones/$id/leida');
    _fijar(_noLeidas > 0 ? _noLeidas - 1 : 0);
    return NotificacionOut.desdeJson(respuesta as Map<String, dynamic>);
  }

  Future<void> leerTodas() async {
    await api.post('/notificaciones/leer-todas');
    _fijar(0);
  }

  @override
  void dispose() {
    _auth.removeListener(_alCambiarSesion);
    _temporizador?.cancel();
    super.dispose();
  }
}
